include/smack/smack.h

```diff
@@ -135,32 +224,143 @@ void __SMACK_decls() {
   D("axiom (forall i: int :: $fp2si($si2fp(i)) == i);");
   D("axiom (forall f: float :: $si2fp($fp2si(f)) == f);");
 
-  // Memory Model
-  D("function $base(int) returns (int);");
+#ifdef BITVECTOR
+  D("type i8 = bv8;");
+  D("type i16 = bv16;");
+  D("type i32 = bv32;");
+  D("type i64 = bv32;");
+  D("type ref = i64;");
+  D("type size_t = i32;");
+#else
+  D("type i8 = int;");
+  D("type i16 = int;");
+  D("type i32 = int;");
+  D("type i64 = int;");
+  D("type ref = int;");
+  D("type size_t = int;");
+#endif
 
-  D("const unique $NULL: int;");
-  D("axiom $NULL == 0;");
+  // Memory Model
   D("const $UNDEF: int;");
+  D("function $base(ref) returns (ref);");
+  D("const unique $NULL: ref;");
+#ifdef BITVECTOR
+  D("axiom $NULL == 0bv32;");
+  D("function {:inline} $pa(pointer: ref, index: ref, size: ref) returns (ref) {$add.i32(pointer, $mul.i32(index, size))}");
```

Comment:
To handle address calculation correctly we should rethink the `$pa` function. The problem is that the `index` and `size` arguments as given in LLVM's `getelementptr` instruction are not generally pointer types — they can be arbitrary types and indexes. If we wanted to handle this correctly according to our current design, we'd have to provide a `$pa.X.Y` function for each type of `index` and `size`.

However, there seems to be a better way. Instead of encoding pointer arithmetic with this `$pa` function, we could encode it via a function `$add.ref(p1: ref, p2: ref) returns (ref)`. All pointer arithmetic should then use this function. This means that our translation of `getelementpointer` will have to change a bit, generating something like
````
$add.ref(p, $zext.i32.ref( $add.i32(i,j) ) )
````
instead of `$pa(p, i, j)`, where extensions via `$zext` should be added when necessary to extend smaller integers to bigger ones. **Note** this might even be necessary inside the `$add`, when the size of `i` and `j` are different. I think this should be `$zext` and not `$sext` since the indices are positive integers, but we should verify this too.